<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>micom.workflows package &#8212; micom 0.16.0 documentation</title>
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootswatch-3.3.4/yeti/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script src="_static/js/jquery-1.11.0.min.js"></script>
    <script src="_static/js/jquery-fix.js"></script>
    <script src="_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script src="_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="micom.viz package" href="micom.viz.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head><body>

<div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
  <div class="container">
    <div class="navbar-header">
      <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">
        micom</a>
      <span class="navbar-text navbar-version pull-left"><b>0.16.x</b></span>
    </div>

      <div class="collapse navbar-collapse nav-collapse">
        <ul class="nav navbar-nav">
          
          
          
          
        </ul>

        
          
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
        
        
          
            <ul class='nav navbar-nav navbar-right'>
                
  <li>
    <a href="micom.viz.html" title="Previous Chapter: micom.viz package"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; micom.viz package</span>
    </a>
  </li>
            </ul>
          
        
      </div>
  </div>
</div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="logic.html">Methods used by MICOM</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing.html">Installing micom</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing.html#Solvers">Solvers</a></li>
<li class="toctree-l1"><a class="reference internal" href="high_level.html">The High Level API</a></li>
<li class="toctree-l1"><a class="reference internal" href="community.html">Building communities</a></li>
<li class="toctree-l1"><a class="reference internal" href="growth_fluxes.html">Growth rates and fluxes</a></li>
<li class="toctree-l1"><a class="reference internal" href="media.html">Growth media</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxa_knockouts.html">Knockouts</a></li>
<li class="toctree-l1"><a class="reference internal" href="elasticities.html">Intervention studies</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Analyzing many samples in parallel</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="micom.html">API</a></li>
</ul>

        </div>
      </div>
    <div class="col-md-9 content">
      
  <div class="section" id="micom-workflows-package">
<h1>micom.workflows package<a class="headerlink" href="#micom-workflows-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-micom.workflows.build">
<span id="micom-workflows-build-module"></span><h2>micom.workflows.build module<a class="headerlink" href="#module-micom.workflows.build" title="Permalink to this headline">¶</a></h2>
<p>Worflow to build models for several samples.</p>
<dl class="py function">
<dt id="micom.workflows.build.build">
<code class="sig-prename descclassname">micom.workflows.build.</code><code class="sig-name descname">build</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">model_db</span></em>, <em class="sig-param"><span class="n">out_folder</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">0.0001</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">solver</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.build.build" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds a series of community models.</p>
<p>This is a best-practice implementation of building community models
for several samples in parallel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>taxonomy</strong> (<em>pandas.DataFrame</em>) – The taxonomy used for building the model. Must have at least the
columns “id” and “sample_id”. This must also
contain at least a column with the same name as the rank used in
the model database. Thus, for a genus-level database you will need
a column <cite>genus</cite>. Additional taxa ranks can also be specified and
will be used to be more stringent in taxa matching.
Finally, the taxonomy should contain a column <cite>abundance</cite>. It will
be used to quantify each individual in the community. If absent,
MICOM will assume all individuals are present in the same amount.</p></li>
<li><p><strong>model_db</strong> (<em>str</em>) – A pre-built model database. If ending in <cite>.qza</cite> must be a Qiime 2
artifact of type <cite>MetabolicModels[JSON]</cite>. Can also be a folder,
zip (must end in <cite>.zip</cite>) file or None if the taxonomy contains a
column <cite>file</cite>.</p></li>
<li><p><strong>out_folder</strong> (<em>str</em>) – The built models and a manifest file will be written to this
folder.</p></li>
<li><p><strong>cutoff</strong> (<em>float in</em><em> [</em><em>0.0</em><em>, </em><em>1.0</em><em>]</em>) – Abundance cutoff. Taxa with a relative abundance smaller than this
will not be included in the model.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
<li><p><strong>solver</strong> (<em>str</em>) – Name of the solver used for the linear and quadratic problems.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The manifest for the built models. Contains taxa abundances,
build metrics and file basenames.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.build.build_and_save">
<code class="sig-prename descclassname">micom.workflows.build.</code><code class="sig-name descname">build_and_save</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">args</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.build.build_and_save" title="Permalink to this definition">¶</a></dt>
<dd><p>Build a single community model.</p>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.build.build_database">
<code class="sig-prename descclassname">micom.workflows.build.</code><code class="sig-name descname">build_database</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">out_path</span></em>, <em class="sig-param"><span class="n">rank</span><span class="o">=</span><span class="default_value">'genus'</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">compress</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">progress</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.build.build_database" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a model database from a set of SBML files.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A manifest for the joined models will also be written to the output folder
as “manifest.csv”. This may contain NA entries for additional columns
that had different values within the summarized models.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – A manifest of SBML files containing their filepath as well as taxonomy.
Must contain the columns “file”, “kingdom”, “phylum”, “class”,
“order”, “family”, “genus”, and “species”. May contain additional
columns.</p></li>
<li><p><strong>out_path</strong> (<em>str</em>) – The directory where the joined models will be written.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
<li><p><strong>compress</strong> (<em>bool</em>) – Whether to compress the output. Default is True if out_path ends with
“.zip” otherwise no.</p></li>
<li><p><strong>progress</strong> (<em>bool</em>) – Whether to show a progress bar.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The manifest of the joined models. Will still contain information
from the original metadata.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-micom.workflows.core">
<span id="micom-workflows-core-module"></span><h2>micom.workflows.core module<a class="headerlink" href="#module-micom.workflows.core" title="Permalink to this headline">¶</a></h2>
<p>Makes it easier to run analyses on several samples in parallel.</p>
<dl class="py class">
<dt id="micom.workflows.core.GrowthResults">
<em class="property">class </em><code class="sig-prename descclassname">micom.workflows.core.</code><code class="sig-name descname">GrowthResults</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">growth_rates</span></em>, <em class="sig-param"><span class="n">exchanges</span></em>, <em class="sig-param"><span class="n">annotations</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.core.GrowthResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="py attribute">
<dt id="micom.workflows.core.GrowthResults.annotations">
<code class="sig-name descname">annotations</code><a class="headerlink" href="#micom.workflows.core.GrowthResults.annotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py attribute">
<dt id="micom.workflows.core.GrowthResults.exchanges">
<code class="sig-name descname">exchanges</code><a class="headerlink" href="#micom.workflows.core.GrowthResults.exchanges" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py attribute">
<dt id="micom.workflows.core.GrowthResults.growth_rates">
<code class="sig-name descname">growth_rates</code><a class="headerlink" href="#micom.workflows.core.GrowthResults.growth_rates" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="micom.workflows.core.workflow">
<code class="sig-prename descclassname">micom.workflows.core.</code><code class="sig-name descname">workflow</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">func</span></em>, <em class="sig-param"><span class="n">args</span></em>, <em class="sig-param"><span class="n">n_jobs</span><span class="o">=</span><span class="default_value">4</span></em>, <em class="sig-param"><span class="n">unit</span><span class="o">=</span><span class="default_value">'sample(s)'</span></em>, <em class="sig-param"><span class="n">progress</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.core.workflow" title="Permalink to this definition">¶</a></dt>
<dd><p>Run analyses for several samples in parallel.</p>
<p>This will analyze several samples in parallel. Includes a workaround for
optlang memory leak.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<em>function</em>) – A function that takes a single argument (can be any object) and
that performs your analysis for a single sample.</p></li>
<li><p><strong>args</strong> (<em>array-like object</em>) – An array-like object (list, tuple, numpy array, pandas Series, etc.)
that contains the arguments for each sample.</p></li>
<li><p><strong>n_jobs</strong> (<em>positive int</em>) – How many samples to analyze in parallel at once.</p></li>
<li><p><strong>unit</strong> (<em>str</em>) – The unit used for the progress bar.</p></li>
<li><p><strong>progress</strong> (<em>bool</em>) – Whether to show a progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-micom.workflows.grow">
<span id="micom-workflows-grow-module"></span><h2>micom.workflows.grow module<a class="headerlink" href="#module-micom.workflows.grow" title="Permalink to this headline">¶</a></h2>
<p>Performs growth and exchange analysis for several models.</p>
<dl class="py function">
<dt id="micom.workflows.grow.grow">
<code class="sig-prename descclassname">micom.workflows.grow.</code><code class="sig-name descname">grow</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">medium</span></em>, <em class="sig-param"><span class="n">tradeoff</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.grow.grow" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate growth for a set of community models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – The manifest as returned by the <cite>build</cite> workflow.</p></li>
<li><p><strong>model_folder</strong> (<em>str</em>) – The folder in which to find the files mentioned in the manifest.</p></li>
<li><p><strong>medium</strong> (<em>pandas.DataFrame</em>) – A growth medium. Must have columns “reaction” and “flux” denoting
exchnage reactions and their respective maximum flux.</p></li>
<li><p><strong>tradeoff</strong> (<em>float in</em><em> (</em><em>0.0</em><em>, </em><em>1.0</em><em>]</em>) – A tradeoff value. Can be chosen by running the <cite>tradeoff</cite> workflow or
by experince. Tradeoff values of 0.5 for metagenomcis data and 0.3 for
16S data seem to work well.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A named tuple containing the growth rates and exchange fluxes for all
samples/models.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#micom.workflows.core.GrowthResults" title="micom.workflows.core.GrowthResults">GrowthResults</a></p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-micom.workflows.media">
<span id="micom-workflows-media-module"></span><h2>micom.workflows.media module<a class="headerlink" href="#module-micom.workflows.media" title="Permalink to this headline">¶</a></h2>
<p>Example workflows for micom.</p>
<dl class="py function">
<dt id="micom.workflows.media.fix_medium">
<code class="sig-prename descclassname">micom.workflows.media.</code><code class="sig-name descname">fix_medium</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">medium</span></em>, <em class="sig-param"><span class="n">min_growth</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">max_import</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">minimize_components</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">summarize</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.media.fix_medium" title="Permalink to this definition">¶</a></dt>
<dd><p>Augment a growth medium so all community members can grow in it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – The manifest as returned by the <cite>build</cite> workflow.</p></li>
<li><p><strong>model_folder</strong> (<em>str</em>) – The folder in which to find the files mentioned in the manifest.</p></li>
<li><p><strong>medium</strong> (<em>pandas.Series</em><em> or </em><em>pandas.DataFrame</em>) – A growth medium with exchange reaction IDs as index and positive
import fluxes as values. If a DataFrame needs columns <cite>flux</cite> and
<cite>reaction</cite>.</p></li>
<li><p><strong>min_growth</strong> (<em>positive float</em>) – The minimum biomass production required for growth.</p></li>
<li><p><strong>max_import</strong> (<em>positive float</em>) – The maximum import rate for added imports.</p></li>
<li><p><strong>minimize_components</strong> (<em>boolean</em>) – Whether to minimize the number of media components rather than the
total flux.</p></li>
<li><p><strong>summarize</strong> (<em>boolean</em>) – Whether to summarize the medium across all samples. If False will
return a medium for each sample.</p></li>
<li><p><strong>threads</strong> (<em>int</em>) – The number of processes to use.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new growth medium with the smallest amount of augmentations such
that all members of the community can grow in it.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.media.minimal_media">
<code class="sig-prename descclassname">micom.workflows.media.</code><code class="sig-name descname">minimal_media</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">summarize</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">min_growth</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.media.minimal_media" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the minimal medium for a set of community models.</p>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.media.process_medium">
<code class="sig-prename descclassname">micom.workflows.media.</code><code class="sig-name descname">process_medium</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">medium</span></em>, <em class="sig-param"><span class="n">samples</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.media.process_medium" title="Permalink to this definition">¶</a></dt>
<dd><p>Prepare a medium for simulation.</p>
</dd></dl>

</div>
<div class="section" id="module-micom.workflows.tradeoff">
<span id="micom-workflows-tradeoff-module"></span><h2>micom.workflows.tradeoff module<a class="headerlink" href="#module-micom.workflows.tradeoff" title="Permalink to this headline">¶</a></h2>
<p>Workflow to run cooperative tradeoff with various tradeoff values.</p>
<dl class="py function">
<dt id="micom.workflows.tradeoff.tradeoff">
<code class="sig-prename descclassname">micom.workflows.tradeoff.</code><code class="sig-name descname">tradeoff</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">medium</span></em>, <em class="sig-param"><span class="n">tradeoffs</span><span class="o">=</span><span class="default_value">array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0])</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.tradeoff.tradeoff" title="Permalink to this definition">¶</a></dt>
<dd><p>Run growth rate predictions for varying tradeoff values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – The manifest as returned by the <cite>build</cite> workflow.</p></li>
<li><p><strong>model_folder</strong> (<em>str</em>) – The folder in which to find the files mentioned in the manifest.</p></li>
<li><p><strong>medium</strong> (<em>pandas.DataFrame</em>) – A growth medium. Must have columns “reaction” and “flux” denoting
exchnage reactions and their respective maximum flux.</p></li>
<li><p><strong>tradeoffs</strong> (<em>array of floats in</em><em> (</em><em>0.0</em><em>, </em><em>1.0</em><em>]</em>) – An array of tradeoff vaues to be tested. One simulation without
a tradeoff (no cooperative tradeoff) will always be run additionally
and will have a tradeoff of “NaN”.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The predicted growth rates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-micom.workflows">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-micom.workflows" title="Permalink to this headline">¶</a></h2>
<p>Init file for MICOM workflows.</p>
<dl class="py function">
<dt id="micom.workflows.workflow">
<code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">workflow</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">func</span></em>, <em class="sig-param"><span class="n">args</span></em>, <em class="sig-param"><span class="n">n_jobs</span><span class="o">=</span><span class="default_value">4</span></em>, <em class="sig-param"><span class="n">unit</span><span class="o">=</span><span class="default_value">'sample(s)'</span></em>, <em class="sig-param"><span class="n">progress</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.workflow" title="Permalink to this definition">¶</a></dt>
<dd><p>Run analyses for several samples in parallel.</p>
<p>This will analyze several samples in parallel. Includes a workaround for
optlang memory leak.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>func</strong> (<em>function</em>) – A function that takes a single argument (can be any object) and
that performs your analysis for a single sample.</p></li>
<li><p><strong>args</strong> (<em>array-like object</em>) – An array-like object (list, tuple, numpy array, pandas Series, etc.)
that contains the arguments for each sample.</p></li>
<li><p><strong>n_jobs</strong> (<em>positive int</em>) – How many samples to analyze in parallel at once.</p></li>
<li><p><strong>unit</strong> (<em>str</em>) – The unit used for the progress bar.</p></li>
<li><p><strong>progress</strong> (<em>bool</em>) – Whether to show a progress bar.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.build">
<code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">build</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">taxonomy</span></em>, <em class="sig-param"><span class="n">model_db</span></em>, <em class="sig-param"><span class="n">out_folder</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">0.0001</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">solver</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.build" title="Permalink to this definition">¶</a></dt>
<dd><p>Builds a series of community models.</p>
<p>This is a best-practice implementation of building community models
for several samples in parallel.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>taxonomy</strong> (<em>pandas.DataFrame</em>) – The taxonomy used for building the model. Must have at least the
columns “id” and “sample_id”. This must also
contain at least a column with the same name as the rank used in
the model database. Thus, for a genus-level database you will need
a column <cite>genus</cite>. Additional taxa ranks can also be specified and
will be used to be more stringent in taxa matching.
Finally, the taxonomy should contain a column <cite>abundance</cite>. It will
be used to quantify each individual in the community. If absent,
MICOM will assume all individuals are present in the same amount.</p></li>
<li><p><strong>model_db</strong> (<em>str</em>) – A pre-built model database. If ending in <cite>.qza</cite> must be a Qiime 2
artifact of type <cite>MetabolicModels[JSON]</cite>. Can also be a folder,
zip (must end in <cite>.zip</cite>) file or None if the taxonomy contains a
column <cite>file</cite>.</p></li>
<li><p><strong>out_folder</strong> (<em>str</em>) – The built models and a manifest file will be written to this
folder.</p></li>
<li><p><strong>cutoff</strong> (<em>float in</em><em> [</em><em>0.0</em><em>, </em><em>1.0</em><em>]</em>) – Abundance cutoff. Taxa with a relative abundance smaller than this
will not be included in the model.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
<li><p><strong>solver</strong> (<em>str</em>) – Name of the solver used for the linear and quadratic problems.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The manifest for the built models. Contains taxa abundances,
build metrics and file basenames.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.build_database">
<code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">build_database</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">out_path</span></em>, <em class="sig-param"><span class="n">rank</span><span class="o">=</span><span class="default_value">'genus'</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">compress</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">progress</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.build_database" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a model database from a set of SBML files.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A manifest for the joined models will also be written to the output folder
as “manifest.csv”. This may contain NA entries for additional columns
that had different values within the summarized models.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – A manifest of SBML files containing their filepath as well as taxonomy.
Must contain the columns “file”, “kingdom”, “phylum”, “class”,
“order”, “family”, “genus”, and “species”. May contain additional
columns.</p></li>
<li><p><strong>out_path</strong> (<em>str</em>) – The directory where the joined models will be written.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
<li><p><strong>compress</strong> (<em>bool</em>) – Whether to compress the output. Default is True if out_path ends with
“.zip” otherwise no.</p></li>
<li><p><strong>progress</strong> (<em>bool</em>) – Whether to show a progress bar.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The manifest of the joined models. Will still contain information
from the original metadata.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.grow">
<code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">grow</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">medium</span></em>, <em class="sig-param"><span class="n">tradeoff</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.grow" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate growth for a set of community models.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – The manifest as returned by the <cite>build</cite> workflow.</p></li>
<li><p><strong>model_folder</strong> (<em>str</em>) – The folder in which to find the files mentioned in the manifest.</p></li>
<li><p><strong>medium</strong> (<em>pandas.DataFrame</em>) – A growth medium. Must have columns “reaction” and “flux” denoting
exchnage reactions and their respective maximum flux.</p></li>
<li><p><strong>tradeoff</strong> (<em>float in</em><em> (</em><em>0.0</em><em>, </em><em>1.0</em><em>]</em>) – A tradeoff value. Can be chosen by running the <cite>tradeoff</cite> workflow or
by experince. Tradeoff values of 0.5 for metagenomcis data and 0.3 for
16S data seem to work well.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A named tuple containing the growth rates and exchange fluxes for all
samples/models.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><a class="reference internal" href="#micom.workflows.core.GrowthResults" title="micom.workflows.core.GrowthResults">GrowthResults</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.tradeoff">
<code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">tradeoff</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">medium</span></em>, <em class="sig-param"><span class="n">tradeoffs</span><span class="o">=</span><span class="default_value">array([0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0])</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.tradeoff" title="Permalink to this definition">¶</a></dt>
<dd><p>Run growth rate predictions for varying tradeoff values.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – The manifest as returned by the <cite>build</cite> workflow.</p></li>
<li><p><strong>model_folder</strong> (<em>str</em>) – The folder in which to find the files mentioned in the manifest.</p></li>
<li><p><strong>medium</strong> (<em>pandas.DataFrame</em>) – A growth medium. Must have columns “reaction” and “flux” denoting
exchnage reactions and their respective maximum flux.</p></li>
<li><p><strong>tradeoffs</strong> (<em>array of floats in</em><em> (</em><em>0.0</em><em>, </em><em>1.0</em><em>]</em>) – An array of tradeoff vaues to be tested. One simulation without
a tradeoff (no cooperative tradeoff) will always be run additionally
and will have a tradeoff of “NaN”.</p></li>
<li><p><strong>threads</strong> (<em>int &gt;=1</em>) – The number of parallel workers to use when building models. As a
rule of thumb you will need around 1GB of RAM for each thread.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>The predicted growth rates.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.fix_medium">
<code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">fix_medium</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">medium</span></em>, <em class="sig-param"><span class="n">min_growth</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">max_import</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">minimize_components</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">summarize</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.fix_medium" title="Permalink to this definition">¶</a></dt>
<dd><p>Augment a growth medium so all community members can grow in it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>manifest</strong> (<em>pandas.DataFrame</em>) – The manifest as returned by the <cite>build</cite> workflow.</p></li>
<li><p><strong>model_folder</strong> (<em>str</em>) – The folder in which to find the files mentioned in the manifest.</p></li>
<li><p><strong>medium</strong> (<em>pandas.Series</em><em> or </em><em>pandas.DataFrame</em>) – A growth medium with exchange reaction IDs as index and positive
import fluxes as values. If a DataFrame needs columns <cite>flux</cite> and
<cite>reaction</cite>.</p></li>
<li><p><strong>min_growth</strong> (<em>positive float</em>) – The minimum biomass production required for growth.</p></li>
<li><p><strong>max_import</strong> (<em>positive float</em>) – The maximum import rate for added imports.</p></li>
<li><p><strong>minimize_components</strong> (<em>boolean</em>) – Whether to minimize the number of media components rather than the
total flux.</p></li>
<li><p><strong>summarize</strong> (<em>boolean</em>) – Whether to summarize the medium across all samples. If False will
return a medium for each sample.</p></li>
<li><p><strong>threads</strong> (<em>int</em>) – The number of processes to use.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A new growth medium with the smallest amount of augmentations such
that all members of the community can grow in it.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="micom.workflows.minimal_media">
<code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">minimal_media</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">manifest</span></em>, <em class="sig-param"><span class="n">model_folder</span></em>, <em class="sig-param"><span class="n">summarize</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">min_growth</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">threads</span><span class="o">=</span><span class="default_value">1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.minimal_media" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the minimal medium for a set of community models.</p>
</dd></dl>

<dl class="py class">
<dt id="micom.workflows.GrowthResults">
<em class="property">class </em><code class="sig-prename descclassname">micom.workflows.</code><code class="sig-name descname">GrowthResults</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">growth_rates</span></em>, <em class="sig-param"><span class="n">exchanges</span></em>, <em class="sig-param"><span class="n">annotations</span></em><span class="sig-paren">)</span><a class="headerlink" href="#micom.workflows.GrowthResults" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tuple</span></code></p>
<dl class="py attribute">
<dt id="micom.workflows.GrowthResults.annotations">
<code class="sig-name descname">annotations</code><a class="headerlink" href="#micom.workflows.GrowthResults.annotations" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 2</p>
</dd></dl>

<dl class="py attribute">
<dt id="micom.workflows.GrowthResults.exchanges">
<code class="sig-name descname">exchanges</code><a class="headerlink" href="#micom.workflows.GrowthResults.exchanges" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 1</p>
</dd></dl>

<dl class="py attribute">
<dt id="micom.workflows.GrowthResults.growth_rates">
<code class="sig-name descname">growth_rates</code><a class="headerlink" href="#micom.workflows.GrowthResults.growth_rates" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for field number 0</p>
</dd></dl>

</dd></dl>

</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017-2020, Christian Diener.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>