<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <title>Visualizations &#8212; micom 0.18.4 documentation</title>
    <link rel="stylesheet" href="_static/basic.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootswatch-3.3.4/yeti/bootstrap.min.css" type="text/css" />
    <link rel="stylesheet" href="_static/bootstrap-sphinx.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script async="async" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/x-mathjax-config">MathJax.Hub.Config({"tex2jax": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true, "ignoreClass": "document", "processClass": "math|output_area"}})</script>
    <script src="_static/js/jquery-1.11.0.min.js"></script>
    <script src="_static/js/jquery-fix.js"></script>
    <script src="_static/bootstrap-3.3.4/js/bootstrap.min.js"></script>
    <script src="_static/bootstrap-sphinx.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Single Model API" href="community.html" />
    <link rel="prev" title="MICOM workflows" href="high_level.html" />
<meta charset='utf-8'>
<meta http-equiv='X-UA-Compatible' content='IE=edge,chrome=1'>
<meta name='viewport' content='width=device-width, initial-scale=1.0, maximum-scale=1'>
<meta name="apple-mobile-web-app-capable" content="yes">

  </head><body>

<div id="navbar" class="navbar navbar-inverse navbar-default navbar-fixed-top">
  <div class="container">
    <div class="navbar-header">
      <!-- .btn-navbar is used as the toggle for collapsed navbar content -->
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".nav-collapse">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">
        micom</a>
      <span class="navbar-text navbar-version pull-left"><b>0.18.x</b></span>
    </div>

      <div class="collapse navbar-collapse nav-collapse">
        <ul class="nav navbar-nav">
          
          
          
          
        </ul>

        
          
<form class="navbar-form navbar-right" action="search.html" method="get">
 <div class="form-group">
  <input type="text" name="q" class="form-control" placeholder="Search" />
 </div>
  <input type="hidden" name="check_keywords" value="yes" />
  <input type="hidden" name="area" value="default" />
</form>
        
        
          
            <ul class='nav navbar-nav navbar-right'>
                
  <li>
    <a href="high_level.html" title="Previous Chapter: MICOM workflows"><span class="glyphicon glyphicon-chevron-left visible-sm"></span><span class="hidden-sm hidden-tablet">&laquo; MICOM workflows</span>
    </a>
  </li>
  <li>
    <a href="community.html" title="Next Chapter: Single Model API"><span class="glyphicon glyphicon-chevron-right visible-sm"></span><span class="hidden-sm hidden-tablet">Single Model API &raquo;</span>
    </a>
  </li>
            </ul>
          
        
      </div>
  </div>
</div>

<div class="container">
  <div class="row">
      <div class="col-md-3">
        <div id="sidebar" class="bs-sidenav" role="complementary"><ul class="current">
<li class="toctree-l1"><a class="reference internal" href="logic.html">Methods used by MICOM</a></li>
<li class="toctree-l1"><a class="reference internal" href="installing.html">Installing MICOM</a></li>
<li class="toctree-l1"><a class="reference internal" href="high_level.html">MICOM workflows</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Visualizations</a></li>
<li class="toctree-l1"><a class="reference internal" href="community.html">Building communities</a></li>
<li class="toctree-l1"><a class="reference internal" href="growth_fluxes.html">Growth rates and fluxes</a></li>
<li class="toctree-l1"><a class="reference internal" href="media.html">Growth media</a></li>
<li class="toctree-l1"><a class="reference internal" href="taxa_knockouts.html">Knockouts</a></li>
<li class="toctree-l1"><a class="reference internal" href="elasticities.html">Intervention studies</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflows.html">Analyzing many samples in parallel</a></li>
<li class="toctree-l1"><a class="reference internal" href="micom.html">API</a></li>
</ul>

        </div>
      </div>
    <div class="col-md-9 content">
      
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container,
div.nbinput.container div.prompt,
div.nbinput.container div.input_area,
div.nbinput.container div[class*=highlight],
div.nbinput.container div[class*=highlight] pre,
div.nboutput.container,
div.nboutput.container div.prompt,
div.nboutput.container div.output_area,
div.nboutput.container div[class*=highlight],
div.nboutput.container div[class*=highlight] pre {
    background: none;
    border: none;
    padding: 0 0;
    margin: 0;
    box-shadow: none;
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    min-width: 5.5ex;
    padding-top: 0.3rem;
    padding-right: 0.3rem;
    text-align: right;
    flex: 0;
}
@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    background: #f5f5f5;
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 0.3rem;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt a.copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
div.rendered_html th {
  font-weight: bold;
}
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}
</style>
<div class="section" id="Visualizations">
<h1>Visualizations<a class="headerlink" href="#Visualizations" title="Permalink to this headline">¶</a></h1>
<p>MICOM provides a a set of visualizations that can be used with the outputs from MICOM workflows. Those visualizations are the same as provided by the <a class="reference external" href="https://github.com/micom-dev/q2-micom">MICOM Qiime 2 plugin</a> but are delivered as single HTML files that bundles interactive graphics and raw data.</p>
<p>To create some more interesting figures here we will use a realistic example data set which is the output of running the MICOM grow workflow on a data set of 8 healthy fecal samples and 8 fecal samples from individuals with colorectal cancer taken from <a class="reference external" href="https://doi.org/10.1158/1940-6207.CAPR-14-0129">https://doi.org/10.1158/1940-6207.CAPR-14-0129</a>. To see the interactive evrsion of a visualization you can click on the provided previews. All visualization contain download buttons to download the raw data used to generate the plot.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">micom.data</span> <span class="k">as</span> <span class="nn">mmd</span>

<span class="n">crc_results</span> <span class="o">=</span> <span class="n">mmd</span><span class="o">.</span><span class="n">crc_results</span><span class="p">()</span>
<span class="n">crc_tradeoff</span> <span class="o">=</span> <span class="n">mmd</span><span class="o">.</span><span class="n">crc_tradeoff</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="section" id="Choosing-a-tradeoff-value">
<h2>Choosing a tradeoff value<a class="headerlink" href="#Choosing-a-tradeoff-value" title="Permalink to this headline">¶</a></h2>
<p>In the original MICOM publication we chose the tradeoff based on comparisons with <em>in vivo</em> replication rates derived from metagenome data. However, we observed that the highest correlation with replication rates is usually achieved at the largest tradeoff value that allows the majority of the taxa to grow. Thus, we can run cooperative tradeoff with varying tradeoff values and look for the characteristic elbow where the majority of the community can grow. This can be done by using the
<code class="docutils literal notranslate"><span class="pre">plot_tradeoff</span></code> function.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">micom.viz</span> <span class="kn">import</span> <span class="n">plot_tradeoff</span>

<span class="n">pl</span> <span class="o">=</span> <span class="n">plot_tradeoff</span><span class="p">(</span><span class="n">crc_tradeoff</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;tradeoff.html&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The returned object is a <code class="docutils literal notranslate"><span class="pre">Visualization</span></code> object that contains the raw data in the data attribute.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pl</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;micom.viz.core.Visualization at 0x7ff4213d78b0&gt;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="n">pl</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
dict_keys([&#39;tradeoff&#39;])
</pre></div></div>
</div>
<p>You could open the visualization in your browser with <code class="docutils literal notranslate"><span class="pre">pl.view()</span></code>. Alternatively you can just open the generated HTML file which would give you something like this:</p>
<p><a class="reference external" href="_static/fit.html"><img alt="image0" src="_images/fit.png" /></a></p>
</div>
<div class="section" id="Plotting-growth-rates">
<h2>Plotting growth rates<a class="headerlink" href="#Plotting-growth-rates" title="Permalink to this headline">¶</a></h2>
<p>The first thing we may want to investigate are the growth rates predicted by MICOM. This can be done with the <code class="docutils literal notranslate"><span class="pre">plot_growth</span></code> function.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">micom.viz</span> <span class="kn">import</span> <span class="n">plot_growth</span>

<span class="n">pl</span> <span class="o">=</span> <span class="n">plot_growth</span><span class="p">(</span><span class="n">crc_results</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;growth_rates.html&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Which will give you the following:</p>
<p><a class="reference external" href="_static/fit.html"><img alt="image0" src="_images/fit.png" /></a></p>
</div>
<div class="section" id="Plotting-consumed-metabolites">
<h2>Plotting consumed metabolites<a class="headerlink" href="#Plotting-consumed-metabolites" title="Permalink to this headline">¶</a></h2>
<p>To get an overview which metabolites are consumed by the entire microbiota we can use the <code class="docutils literal notranslate"><span class="pre">plot_exchanges_per_sample</span></code> function.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">micom.viz</span> <span class="kn">import</span> <span class="n">plot_exchanges_per_sample</span>

<span class="n">pl</span> <span class="o">=</span> <span class="n">plot_exchanges_per_sample</span><span class="p">(</span><span class="n">crc_results</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;consumption.html&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>This will give you a heatmap showing all consumed components. Unless specified otherwise in the function arguments samples will be clustered so that samples with similar consumption profiles will be close.</p>
<p><a class="reference external" href="_static/fit.html"><img alt="image0" src="_images/fit.png" /></a></p>
</div>
<div class="section" id="Plotting-growth-niches">
<h2>Plotting growth niches<a class="headerlink" href="#Plotting-growth-niches" title="Permalink to this headline">¶</a></h2>
<p>What is consumed globally may be interesting but we may want to know even more how the available growth niches are occupied by the taxa in the sample. This can be done with <code class="docutils literal notranslate"><span class="pre">plot_exchanges_per_taxon</span></code> which will embed the import fluxes for each taxon into two dimension using TSNE and plot the niche occupation map. Here taxa that overlap compete for similar sets of resources. The center of the map denotes the most competitive niche whereas the outskirts of the map denote more specialized
consumption preferences.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">micom.viz</span> <span class="kn">import</span> <span class="n">plot_exchanges_per_taxon</span>

<span class="n">pl</span> <span class="o">=</span> <span class="n">plot_exchanges_per_taxon</span><span class="p">(</span><span class="n">crc_results</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;niche.html&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p><a class="reference external" href="_static/fit.html"><img alt="image0" src="_images/fit.png" /></a></p>
</div>
<div class="section" id="Investigating-the-relationship-with-the-phenotype">
<h2>Investigating the relationship with the phenotype<a class="headerlink" href="#Investigating-the-relationship-with-the-phenotype" title="Permalink to this headline">¶</a></h2>
<p>Finally we may want to which functions relate to an observed phenotype. This can be done with the <code class="docutils literal notranslate"><span class="pre">plot_fit</span></code> function which will:</p>
<ol class="arabic simple">
<li><p>calculate overall production fluxes (total metabolite amount produced by the microbiota)</p></li>
<li><p>log-transform and standardize production fluxes</p></li>
<li><p>train a LASSO regression (continuous response) or LASSO logistic regression (binary response)</p></li>
<li><p>present the coefficients and differences in production fluxes</p></li>
</ol>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">micom.viz</span> <span class="kn">import</span> <span class="n">plot_fit</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">samples</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">crc_results</span><span class="o">.</span><span class="n">growth_rates</span><span class="o">.</span><span class="n">sample_id</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
<span class="n">has_cancer</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">samples</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;Cancer&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">samples</span><span class="p">)</span>
<span class="n">has_cancer</span>
<span class="n">pl</span> <span class="o">=</span> <span class="n">plot_fit</span><span class="p">(</span><span class="n">crc_results</span><span class="p">,</span> <span class="n">phenotype</span><span class="o">=</span><span class="n">has_cancer</span><span class="p">,</span> <span class="n">variable_type</span><span class="o">=</span><span class="s2">&quot;binary&quot;</span><span class="p">,</span> <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;fit.html&quot;</span><span class="p">,</span> <span class="n">min_coef</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/home/cdiener/code/micom/micom/viz/prediction.py:163: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exchanges[&#34;meta&#34;] = meta[exchanges.sample_id].values
/home/cdiener/code/micom/micom/viz/prediction.py:164: SettingWithCopyWarning:
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  exchanges[&#34;description&#34;] = anns.loc[exchanges.metabolite, &#34;name&#34;].values
</pre></div></div>
</div>
<p>The warnings can be ignored here. The output will look something like the following.</p>
<p><a class="reference external" href="_static/fit.html"><img alt="image0" src="_images/fit.png" /></a></p>
<p>So we see some higher sulfate production in cancer cases but also production of harmful components as TMAO. However, with the low sample numbers we don’t get a very good prediction based on fluxes.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre>
<span></span>
</pre></div>
</div>
</div>
</div>
</div>


    </div>
      
  </div>
</div>
<footer class="footer">
  <div class="container">
    <p class="pull-right">
      <a href="#">Back to top</a>
      
    </p>
    <p>
        &copy; Copyright 2017-2020, Christian Diener.<br/>
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 3.0.1.<br/>
    </p>
  </div>
</footer>
  </body>
</html>